### ISSUE TITLE
element picker cannot peer into embedded YouTube videos

### ISSUE THREAD

**Issue by joey04** [uBlockOrigin/uBlock-issues#103]
This is a follow-on to #102 that I just posted. Same setup. The picker cannot access the "More videos" overlay of a paused video.

I'm posting this separately in case a separate code change would be needed.

This may be a duplicate of https://github.com/gorhill/uBlock/issues/3036 too, but I want to note it here because it's for the picker not the logger inspector.

**Comment by uBlock-user** [comment]
Picker cannot dig into iframes, see - https://github.com/gorhill/uBlock/issues/1744

**Comment by joey04** [comment]
Thanks for finding the duplicate. I was unaware of that one.

For completeness sake, I just confirmed on Pale Moon 27 with an older version of uBO that it can't peer into the iframe of the same scenario.

**Comment by uBlock-user** [comment]
Apparently, it doesn't drill since the beginning, see - `https://github.com/uBlock-LLC/uBlock/issues/88`

**Comment by joey04** [comment]
@gorhill -- Could you briefly explain why the picker cannot drill into iframes? (None of the other issues have an explanation, which is why I'm asking here.)

In particular, I'm wondering if adding this ability would be a can of worms with potentially bad side-effects. Or is it more a matter of a very tricky item to add?

The reason I ask is I came across a different instance of this today, and after uBlock-user pointed out the old duplicates in this thread 2 months ago, I've spotted a few others as well. For me, this is the one open uBO issue that would be a definite capability boost. After all, the convenience of the picker is a very nice luxury. (But if it never gets added, that's not a big deal, so I'm not pleading for it here. I would appreciate a brief explanation, though.)

And once again, I must express gratitude for all that you've done on this excellent tool. It really makes browsing the web so much better!

**Comment by joey04** [comment]
After posting yesterday, I spent some time looking into this and have reached closure on it for myself.

From what I can gather in the code, the use of DOM calls like `document.getBoundingClientRect` and `document.elementFromPoint` may only be valid for the main frame. But I don't know enough about this to be sure, so can't really know how tricky it would be to add iframe support.

Another consideration came to mind: Even if it were simple to implement, it may not be a good idea to do so. Novice users could get confused if the picker transparently drills into an iframe and ends up creating a cosmetic rule for a third-party domain, e.g. a Disqus thread embedded at the end of an article. Perhaps this is not much of a concern, though.

Finally, I'll share what I've learned on how to generate cosmetic rules for subdocument elements:
* One way is to paste the subdoc URL from the logger into a new tab; this loads it as its own main frame, which means the picker will work. (I did this yesterday to create a rule to hide the social icons on embedded Disqus iframes.) But in some cases the desired element may not be visible, so the next step is to launch the browser's page source viewer to try to figure out what the element is. (Last year I wanted a style rule to increase the text size of embedded Instagram comments, so I had to view page source to find the DOM class name.)
* Another way is simply to search the Annoyances lists and see if they have an item. (I'm pretty sure that's how I got the Youtube hiding rule linked in the first post of this issue; I've had that rule for a while and probably copied it from Fanboy Annoyances.)
